# This is my notebook on Chapter 1 from Wooldridge
### I did 3 computer exercises

In [95]:
import wooldridge
import statsmodels.formula.api as smf
import numpy as np
from scipy import stats

import warnings
warnings.filterwarnings('ignore')

## Task1

In [96]:
data = wooldridge.data('VOTE1')
data

state  district  democA  voteA     expendA      expendB  prtystrA  \
0      AL         7       1     68  328.295990     8.737000        41   
1      AK         1       0     62  626.377014   402.476990        60   
2      AZ         2       1     73   99.607002     3.065000        55   
3      AZ         3       0     69  319.690002    26.281000        64   
4      AR         3       0     75  159.220993    60.054001        66   
..    ...       ...     ...    ...         ...          ...       ...   
168    WV         4       0     39   32.039001   152.270996        42   
169    WI         3       1     32   22.625999   359.800995        53   
170    WI         5       0     36  197.460007  1278.526001        36   
171    WI         7       0     38  202.591003   450.716003        46   
172    WI         8       1     30   14.421000   227.822998        47   

     lexpendA  lexpendB     shareA  
0    5.793916  2.167567  97.407669  
1    6.439952  5.997638  60.881039  
2    4.601233  1.120048  97.014763  
3    5.767352  3.268846  92.403702  
4    5.070293  4.095244  72.612473  
..        ...       ...        ...  
168  3.466954  5.025662  17.383217  
169  3.119100  5.885551   5.916420  
170  5.285536  7.153463  13.378174  
171  5.311189  6.110837  31.010078  
172  2.668685  5.428569   5.953087  

[173 rows x 10 columns]

In [97]:
model = smf.ols(data=data, formula='voteA ~ np.log(expendA) + np.log(expendB) + prtystrA')
fitted = model.fit()
print(fitted.summary())

                            OLS Regression Results                            
Dep. Variable:                  voteA   R-squared:                       0.793
Model:                            OLS   Adj. R-squared:                  0.789
Method:                 Least Squares   F-statistic:                     215.2
Date:                Wed, 04 May 2022   Prob (F-statistic):           1.76e-57
Time:                        07:52:24   Log-Likelihood:                -596.86
No. Observations:                 173   AIC:                             1202.
Df Residuals:                     169   BIC:                             1214.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          45.0789      3.926     

#### Now, let's test the hypothesis:
    
$ H_0 : \beta_1 = - \beta_2.$
   
To do this, let's set $ \theta = \beta_1 + \beta_2 $, or $ \beta_2 = \theta - \beta_1 $.

Now our regression will look like:

$voteA = \beta_0 + \beta_1 \cdot [ log(expendA) - log(expendB) ] + \theta_1 \cdot log(expendB) + \beta_3 \cdot prtystrA + u$

In [98]:
df = data[ 'voteA$expendA$expendB$prtystrA'.split('$') ].copy()
df[['expendA', 'expendB']] = df[['expendA', 'expendB']].apply(np.log)
df['log_diff'] = df['expendA'] - df['expendB']

model = smf.ols(data=df, formula='voteA ~ log_diff + expendB + prtystrA')
fitted = model.fit()
print(fitted.summary())

                            OLS Regression Results                            
Dep. Variable:                  voteA   R-squared:                       0.793
Model:                            OLS   Adj. R-squared:                  0.789
Method:                 Least Squares   F-statistic:                     215.2
Date:                Wed, 04 May 2022   Prob (F-statistic):           1.76e-57
Time:                        07:52:24   Log-Likelihood:                -596.86
No. Observations:                 173   AIC:                             1202.
Df Residuals:                     169   BIC:                             1214.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     45.0789      3.926     11.481      0.0

As we can see, $\theta$ is not significant, so we cannot reject the null hypothesis.

## Task 5

In [99]:
data = wooldridge.data('MLB1')
data.head()

salary     teamsal  nl  years  games  atbats  runs  hits  doubles  \
0  6329213.0  38407380.0   1     12   1705    6705  1076  1939      320   
1  3375000.0  38407380.0   1      8    918    3333   407   863      156   
2  3100000.0  38407380.0   1      5    751    2807   370   840      148   
3  2900000.0  38407380.0   1      8   1056    3337   405   816      143   
4  1650000.0  38407380.0   1     12   1196    3603   437   928       19   

   triples  ...     runsyr   percwhte   percblck   perchisp     blckpb  \
0       67  ...  89.666664  70.277969  18.844229  10.877804   0.000000   
1       38  ...  50.875000  70.277969  18.844229  10.877804  18.844229   
2       18  ...  74.000000  70.277969  18.844229  10.877804   0.000000   
3       18  ...  50.625000  70.277969  18.844229  10.877804   0.000000   
4       16  ...  36.416668  70.277969  18.844229  10.877804  18.844229   

   hispph     whtepw     blckph  hisppb    lsalary  
0     0.0  70.277969   0.000000     0.0  15.660686  
1     0.0   0.000000  10.877804     0.0  15.031906  
2     0.0  70.277969   0.000000     0.0  14.946913  
3     0.0  70.277969   0.000000     0.0  14.880221  
4     0.0   0.000000  10.877804     0.0  14.316286  

[5 rows x 47 columns]

In [100]:
model = smf.ols(data=data, formula='np.log(salary) ~ years + gamesyr + bavg + hrunsyr')
fitted = model.fit()
fitted.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:         np.log(salary)   R-squared:                       0.625
Model:                            OLS   Adj. R-squared:                  0.621
Method:                 Least Squares   F-statistic:                     145.2
Date:                Wed, 04 May 2022   Prob (F-statistic):           6.98e-73
Time:                        07:52:24   Log-Likelihood:                -386.25
No. Observations:                 353   AIC:                             782.5
Df Residuals:                     348   BIC:                             801.8
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     11.0209      0.266     41.476      0.000      10.498      11.544
years          0.0677      0.012      5.592      0.000       0.044       0.092
gamesyr        0.0158      0.002     10.079      0.000       0.013       0.019
bavg           0.0014      0.001      1.331      0.184      -0.001       0.004
hrunsyr        0.0359      0.007      4.964      0.000       0.022       0.050
==============================================================================
Omnibus:                        7.369   Durbin-Watson:                   1.244
Prob(Omnibus):                  0.025   Jarque-Bera (JB):               11.240
Skew:                          -0.085   Prob(JB):                      0.00362
Kurtosis:                       3.858   Cond. No.                     1.90e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.9e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [101]:
model2 = smf.ols(data=data, formula='np.log(salary) ~ years + gamesyr + bavg + hrunsyr + \
                                    runsyr + fldperc + sbasesyr')
fitted2 = model2.fit()
fitted2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:         np.log(salary)   R-squared:                       0.639
Model:                            OLS   Adj. R-squared:                  0.632
Method:                 Least Squares   F-statistic:                     87.25
Date:                Wed, 04 May 2022   Prob (F-statistic):           1.84e-72
Time:                        07:52:24   Log-Likelihood:                -379.71
No. Observations:                 353   AIC:                             775.4
Df Residuals:                     345   BIC:                             806.3
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     10.4083      2.003      5.196      0.000       6.468      14.348
years          0.0700      0.012      5.844      0.000       0.046       0.094
gamesyr        0.0079      0.003      2.950      0.003       0.003       0.013
bavg           0.0005      0.001      0.480      0.632      -0.002       0.003
hrunsyr        0.0232      0.009      2.687      0.008       0.006       0.040
runsyr         0.0174      0.005      3.434      0.001       0.007       0.027
fldperc        0.0010      0.002      0.516      0.606      -0.003       0.005
sbasesyr      -0.0064      0.005     -1.238      0.216      -0.017       0.004
==============================================================================
Omnibus:                        1.205   Durbin-Watson:                   1.287
Prob(Omnibus):                  0.548   Jarque-Bera (JB):                0.956
Skew:                           0.090   Prob(JB):                        0.620
Kurtosis:                       3.180   Cond. No.                     5.33e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.33e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Now, let's test single significance hypothesis, let's say on sbasesyr.

$t_{\beta_{sbasesyr}, 353} \approx z = \frac{\beta_{sbasesyr} - 0}{se(\beta_{sbasesyr})}$

In [102]:
critical = stats.t.isf(0.025, 345)
beta, se_beta = fitted.params[-1], fitted.bse[-1]
t_value = (fitted.params[-1]) / se_beta
t_value

4.96399615507859

Not statistically significan't. We cannot find enough evidence to reject null hypothesis.

In [103]:
df = data[['salary', 'bavg', 'fldperc', 'sbasesyr']]
df.head()

salary   bavg  fldperc   sbasesyr
0  6329213.0  289.0      989  26.166666
1  3375000.0  259.0      968  16.625000
2  3100000.0  299.0      994   8.200000
3  2900000.0  245.0      971   1.875000
4  1650000.0  258.0      977   2.666667

In [104]:
model3 = smf.ols(data=df, formula='np.log(salary) ~ bavg + fldperc + sbasesyr')
fitted3 = model3.fit()
print(fitted3.summary())
df['lsalary'] = np.log(df['salary'])
X = df[['bavg', 'fldperc', 'sbasesyr']]
df['predicted'] = fitted3.predict(X)
df.head()

                            OLS Regression Results                            
Dep. Variable:         np.log(salary)   R-squared:                       0.191
Model:                            OLS   Adj. R-squared:                  0.184
Method:                 Least Squares   F-statistic:                     27.48
Date:                Wed, 04 May 2022   Prob (F-statistic):           5.60e-16
Time:                        07:52:24   Log-Likelihood:                -522.12
No. Observations:                 353   AIC:                             1052.
Df Residuals:                     349   BIC:                             1068.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      3.8991      2.931      1.330      0.1

salary   bavg  fldperc   sbasesyr    lsalary  predicted
0  6329213.0  289.0      989  26.166666  15.660686  14.373253
1  3375000.0  259.0      968  16.625000  15.031906  13.689849
2  3100000.0  299.0      994   8.200000  14.946913  13.902068
3  2900000.0  245.0      971   1.875000  14.880221  13.135909
4  1650000.0  258.0      977   2.666667  14.316286  13.300261

In [105]:
SSR_r = ((df['lsalary'] - df['lsalary'].mean()) ** 2).sum()
SSR_r

492.17550663447923

In [106]:
SSR_ur = ((df['lsalary'] - df['predicted']) ** 2).sum()
SSR_ur

398.12146161647115

In [107]:
model.nobs

353.0

In [108]:
f_statistic = ((SSR_r - SSR_ur) / 3) / (SSR_ur / 349)
f'Manual f-statistic {f_statistic}'

'Manual f-statistic 27.48312167347774'

In [109]:
f'statsmodels f-statistic {fitted2.fvalue}'

'statsmodels f-statistic 87.24772866151078'

Now, let's do f_test using builtin method from statsmodels

In [110]:
fitted2.f_test(np.array(((0, 0, 0, 1, 0, 0, 0, 0), (0, 0, 0, 0, 0, 0, 1, 0), (0, 0, 0, 0, 0, 0, 0, 1))))

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=0.685004152289007, p=0.5617088027791233, df_denom=345, df_num=3>

## Task 11

In [111]:
data = wooldridge.data('HTV')
data.head()

wage      abil  educ  ne  nc  west  south  exper  motheduc  fatheduc  \
0  12.019231  5.027738    15   0   0     1      0      9        12        12   
1   8.912656  2.037170    13   1   0     0      0      8        12        10   
2  15.514334  2.475895    15   1   0     0      0     11        12        16   
3  13.333333  3.609240    15   1   0     0      0      6        12        12   
4  11.070110  2.636546    13   1   0     0      0     15        12        15   

   ...  ne18  nc18  south18  west18  urban18    tuit17     tuit18     lwage  \
0  ...     1     0        0       0        1  7.582914   7.260242  2.486508   
1  ...     1     0        0       0        1  8.595144   9.499537  2.187472   
2  ...     1     0        0       0        1  7.311346   7.311346  2.741764   
3  ...     1     0        0       0        1  9.499537  10.162070  2.590267   
4  ...     1     0        0       0        1  7.311346   7.311346  2.404249   

   expersq     ctuit  
0       81 -0.322671  
1       64  0.904392  
2      121  0.000000  
3       36  0.662534  
4      225  0.000000  

[5 rows x 23 columns]

In [112]:
model = smf.ols(data=data, formula='educ ~ motheduc + fatheduc + abil + np.square(abil)')
fitted = model.fit()
fitted.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.444
Model:                            OLS   Adj. R-squared:                  0.443
Method:                 Least Squares   F-statistic:                     244.9
Date:                Wed, 04 May 2022   Prob (F-statistic):          1.34e-154
Time:                        07:52:24   Log-Likelihood:                -2436.6
No. Observations:                1230   AIC:                             4883.
Df Residuals:                    1225   BIC:                             4909.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           8.2402      0.287     28.671      0.000       7.676       8.804
motheduc            0.1901      0.028      6.767      0.000       0.135       0.245
fatheduc            0.1089      0.020      5.558      0.000       0.070       0.147
abil                0.4015      0.030     13.255      0.000       0.342       0.461
np.square(abil)     0.0506      0.008      6.093      0.000       0.034       0.067
==============================================================================
Omnibus:                       45.933   Durbin-Watson:                   1.820
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               56.769
Skew:                           0.404   Prob(JB):                     4.71e-13
Kurtosis:                       3.674   Cond. No.                         115.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Now I'll test hypothesis: $H_0:$ relationship between educ and abil is linear vs $H_1:$ relationship is quadratic.

In [113]:
fitted.f_test(np.array(((0, 0, 0, 1, 0), (0, 0, 0, 0, 1)))).summary()

'<F test: F=215.0602024007355, p=8.917204532088574e-81, df_denom=1.22e+03, df_num=2>'

We see, that relationship is quadratic.

In [114]:
fitted.f_test('(abil=0), (np.square(abil)=0)').summary()

'<F test: F=215.0602024007355, p=8.917204532088574e-81, df_denom=1.22e+03, df_num=2>'

In [115]:
model = smf.ols(data=data, formula='educ ~ motheduc + fatheduc + abil + np.square(abil) + \
                tuit17 + tuit18')
fitted = model.fit()
# fitted.summary()

In [116]:
fitted.f_test('(tuit17=0), (tuit18=0)').summary()

'<F test: F=0.8393288742997933, p=0.432249037805019, df_denom=1.22e+03, df_num=2>'

In [117]:
data['tuit_avg'] = (data['tuit17'] + data['tuit18']) / 2
model = smf.ols(data=data, formula='educ ~ motheduc + fatheduc + abil + np.square(abil) + \
                tuit_avg')
fitted = model.fit()
fitted.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.445
Model:                            OLS   Adj. R-squared:                  0.443
Method:                 Least Squares   F-statistic:                     196.4
Date:                Wed, 04 May 2022   Prob (F-statistic):          9.75e-154
Time:                        07:52:24   Log-Likelihood:                -2435.8
No. Observations:                1230   AIC:                             4884.
Df Residuals:                    1224   BIC:                             4914.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           8.0813      0.313     25.851      0.000       7.468       8.695
motheduc            0.1929      0.028      6.847      0.000       0.138       0.248
fatheduc            0.1084      0.020      5.529      0.000       0.070       0.147
abil                0.3991      0.030     13.156      0.000       0.340       0.459
np.square(abil)     0.0506      0.008      6.095      0.000       0.034       0.067
tuit_avg            0.0160      0.012      1.290      0.197      -0.008       0.040
==============================================================================
Omnibus:                       45.884   Durbin-Watson:                   1.824
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               56.400
Skew:                           0.406   Prob(JB):                     5.66e-13
Kurtosis:                       3.664   Cond. No.                         136.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [118]:
fitted.t_test('(tuit_avg=0)').summary()

<class 'statsmodels.iolib.table.SimpleTable'>

In [119]:
fitted.f_test('(tuit_avg=0)').summary()

'<F test: F=1.6643598346996562, p=0.19725936900521307, df_denom=1.22e+03, df_num=1>'